iFixit/react-commerce
# Vulcan: Add TOC (Desktop/ Mobile) and Scroll Indicator

Some choices made:

* TOC items are stored via their name
	* This means no duplicate names are currently allowed, the last one wins
	* We may want to take a second pass and force them to require ids and for them to be unique, but we found the API may not return ids for all headings right now

CR
---

There was a lot of back and forth after about the TOC logic after 38c2ab6b7540f92bd87d67358d5357d18ce1a6f7.

I would CR commit by commit up to there then start looking at the new files TOC / TOCContext as greenfield and look at the overall diff.

 
QA
---
* Ensure TOC is **off** on the `react-commerce` deployment
* Ensure TOC is **on** on the `react-commerce-prod` deployment
* Ensure TOC behaves well on all screen sizes
* Ensure TOC matches [figma](https://www.figma.com/file/jRG4EyQXRDMGfyLqut08KP/iFixit---Working-files?node-id=13998%3A321530&mode=dev) and has behavior like the [example app](https://rwddn4.csb.app/)
* Here is a [good example of small many toc items](https://react-commerce-prod-git-scroll-add-ui-ifixit.vercel.app/Troubleshooting/Chainsaw/Overheating/484453)
* Here is [a more normal example](https://react-commerce-prod-git-scroll-add-ui-ifixit.vercel.app/Troubleshooting/Television/TV+Has+Sound+But+No+Picture/493422)
* Ensure the Desktop TOC is rendered server-side
	* No flash in
* Ensure the Mobile TOC will pop in once scrolled to
	* This will pop in on refresh
	* Ensure no page shift / CLS occurs
* Ensure Desktop TOC activates the correct item on page load
	* Ensure no URL hash, scroll down and refresh the page so you load at the same scroll position
* Ensure Scroll indicator /TOC are removed if we entirely scroll past content (Mobile mostly, but code doesn't care)
* Ensure selecting an item on the TOC scrolls you to it and that item is selected
	* The code cares about scroll positioning only to highlight so this is checking the math to where to scroll is correct
* Ensure selecting an item is the TOC highlights it briefly
	* Not part of the spec, but I liked it
* Ensure the TOC never selects and item off screen
* Ensure the TOC is scrollable if required
	* Ensure a scroll fade occurs on desktop / mobile TOCs when the TOC is scrollable
	* Ensure no fade if not scrollable
* Ensure the TOC selects the active item as you scroll down the page
* Ensure every possible TOC item is selected when scrolling up / down a page
* Ensure clicking a TOC item updates the URL hash (and browser history) to the new item
	* Test history by pressing the back button

Connects: https://github.com/iFixit/ifixit/issues/49284
